### PR TITLE
Initian design for OTHER status[DRAFT]

### PR DESCRIPTION
**Pull request by hemanttanwar**
This is for design discussion for OTHER status. Test will fail since they are not updated.

### CODE REVIEW

File: core/azure-core/src/main/java/com/azure/core/util/polling/Poller.java [hunk 1]
```diff
@@ -142,15 +143,15 @@ public Poller(Duration pollInterval, Function<PollResponse<T>, Mono<PollResponse
      * The next poll cycle will be defined by retryAfter value in {@link PollResponse}.
      * In absence of {@link PollResponse#getRetryAfter()}, the {@link Poller} will use {@code pollInterval}.
      *
-     * @param pollInterval Not-null and greater than zero poll interval.
-     * @param pollOperation The polling operation to be called by the {@link Poller} instance. This is a callback into the client library,
-     * which must never return {@code null}, and which must always have a non-null {@link com.azure.core.util.polling.PollResponse.OperationStatus}.
-     *{@link Mono} returned from poll operation should never return {@link Mono#error(Throwable)}.If any unexpected scenario happens in poll operation,
+     * @param pollInterval    Not-null and greater than zero poll interval.
+     * @param pollOperation   The polling operation to be called by the {@link Poller} instance. This is a callback into the client library,
```
Comment (JonathanGiles) [inline core/azure-core/src/main/java/com/azure/core/util/polling/Poller.java]: Don't do this kind of alignment

File: core/azure-core/src/main/java/com/azure/core/util/polling/Poller.java [hunk 2]
```diff
@@ -185,12 +186,59 @@ public void cancelOperation() throws UnsupportedOperationException {
...
+        if (observeState == null && (observeOtherStates == null || observeOtherStates.size() == 0)) {
+            throw new IllegalArgumentException("observeOperationStates and observeStates both can not be null or empty.");
+        }
+        //TODO : Design discussion, We may decide to add complete state. If user did not added them, we may never listen them and LRO would never complete.
+        //observeState.add(OperationStatus.State.SUCCESSFULLY_COMPLETED);
+        //observeState.add(OperationStatus.State.USER_CANCELLED);
+        //observeState.add(OperationStatus.State.FAILED);
+        this.fluxHandle = this.fluxHandle.filterWhen(tPollResponse -> matchesState(tPollResponse, observeState, observeOtherStates));
+        return this.fluxHandle;
+    }
```
Comment (JonathanGiles) [inline core/azure-core/src/main/java/com/azure/core/util/polling/Poller.java]: I don't think this API is necessary. The user can implement the filter on their side.

**Comment by JonathanGiles** [comment]
Be sure to refer back to the issue so context can be track - this PR is for issue #3930

File: core/azure-core/src/main/java/com/azure/core/util/polling/Poller.java (hunk 1 above)
Comment (hemanttanwar) [reply core/azure-core/src/main/java/com/azure/core/util/polling/Poller.java]: I  have auto format in Intellij which I should turn off.

File: core/azure-core/src/main/java/com/azure/core/util/polling/Poller.java (hunk 2 above)
Comment (hemanttanwar) [reply core/azure-core/src/main/java/com/azure/core/util/polling/Poller.java]: That's true , makes poller simpler.

File: core/azure-core/src/main/java/com/azure/core/util/polling/PollResponse.java [hunk 3]
```diff
@@ -59,18 +67,31 @@
...
+     * {@code null} is valid value for all other {@link OperationStatus}.
      * @param value The value as a result of poll operation. This can be any custom user-defined object. Null is also valid.
      * @param retryAfter Represents the delay the service has requested until the next polling operation is performed.
      * A {@code null}, zero or negative value will be taken to mean that the {@link Poller} should determine on its own when the next poll operation is to occur.
      * @param properties A map of properties provided by the service that will be made available into the next poll operation.
      * @throws NullPointerException If {@code status} is {@code null}.
+     * @throws IllegalArgumentException if otherStatus is null or empty when status is {@link OperationStatus#OTHER}.
      */
-    public PollResponse(OperationStatus status, T value, Duration retryAfter, Map<Object, Object> properties) {
+    public PollResponse(OperationStatus status, String otherStatus, T value, Duration retryAfter, Map<Object, Object> properties) {
```
Comment (JonathanGiles) [inline core/azure-core/src/main/java/com/azure/core/util/polling/PollResponse.java]: I would simplify the constructor - if the user is supply an 'otherStatus', you don't need them to supply an operation status, you can automatically make that OTHER.

**Pull request closed without merging**